### PR TITLE
Add x86 simd for weighted sse computation

### PR DESCRIPTION
**Pull request by KyleSiefring**

### CODE REVIEW

File: src/x86/sse.asm [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    movq               rax, m1
+    RET
+
+INIT_XMM ssse3
+cglobal weighted_sse_4x16, 6, 7, 7, \
+        src, src_stride, dst, dst_stride, scale, scale_stride, \
+        h
+    mova                m0, [addsub]
+    pxor                m6, m6
+    mov                 hd, 1
```
Comment (cyanreg) [inline src/x86/sse.asm:207]: Just get rid of the loop here. You save a register and a (maybe well) predicted branch.
You can also just macro what's inside the loop and call it twice, would be the same number of lines.

File: src/x86/sse.asm [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+
+%macro WEIGHTED_SSE 2 ; w, h
+%if %1 == 8
+%if %2 == 4
+; Use scale_stride's register to store src_stride3
+cglobal weighted_sse_%1x%2, 6, 7, 5, \
+        src, src_stride, dst, dst_stride, scale, \
+        src_stride3, dst_stride3
+%else
+cglobal weighted_sse_%1x%2, 6, 9, 6, \
```
Comment (cyanreg) [inline src/x86/sse.asm:335]: You can have C-like expressions when declaring the number of registers.
e.g. cglobal weighted_sse_%1x%2, 6, 9 + 2*(%1 == 32 ? 9 >> 1 : 2), 6 + 4*(%2 == 32), and so on.
So you can clean up the ifdeffery somewhat.

File: src/x86/sse.asm [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+; This source code is subject to the terms of the BSD 2 Clause License and
+; the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+; was not distributed with this source code in the LICENSE file, you can
+; obtain it at www.aomedia.org/license/software. If the Alliance for Open
+; Media Patent License 1.0 was not distributed with this source code in the
+; PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+%include "ext/x86/x86inc.asm"
+
+%if ARCH_X86_64
```
Comment (cyanreg) [inline src/x86/sse.asm:12]: Don't think this is needed, most of the functions use less than 8 registers.
You can have a separate WEIGHTED_SSE X, Y if block for those that go above 8.
But no strong opinion on this if it makes the rust code more complicated.

File: src/x86/sse.asm [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+SECTION .text
+
+%define m(x) mangle(private_prefix %+ _ %+ x %+ SUFFIX)
+
+; Consolidate scaling and rounding to one place so that it is easier to change.
+
+%macro SSE_SCALE_4X4 0
+    ; Multiply and shift using scalar code
+    mov             scaled, [scaleq]
+    imul               rax, scaleq
```
[the remark this answers is not in the document]
Comment (KyleSiefring) [reply src/x86/sse.asm:27]: scale is stored as 32 bits so this would load two elements.

File: src/x86/sse.asm (hunk 3 above)
Comment (KyleSiefring) [reply src/x86/sse.asm:12]: I don't think much assembly is hooked up for x86-32, but I'll gate that won't work on x86 for when we get around to it.

File: src/x86/sse.asm (hunk 1 above)
Comment (KyleSiefring) [reply src/x86/sse.asm:207]: Done.

File: src/x86/sse.asm (hunk 2 above)
Comment (cyanreg) [reply src/x86/sse.asm:330]: No, this should work on yasm and ancient nasm versions as well. Its used pretty extensively in ffmpeg.
Regardless, we don't support yasm or any old versions of nasm.

File: src/x86/sse.asm (hunk 4 above)
[the remark this answers is not in the document]
Comment (cyanreg) [reply src/x86/sse.asm:25]: `imul eax, dword [scaleq]` will work then.

[the remark this answers is not in the document]
Comment (KyleSiefring) [reply src/x86/sse.asm:25]: Is there a way to get this to work with 64 bit multiplies?

btw, I'm probably going to change scale to 16 bits at some point anyways.

[the remark this answers is not in the document]
Comment (cyanreg) [reply src/x86/sse.asm:25]: Operands must match, so unfortunately no. Still, 32 bits is a lot of range.

**Pull request merged**